### PR TITLE
feat(protocol-designer): implement ui for flow rate

### PR DESCRIPTION
**Pull request by IanLondon**
## overview

This is UI-only implementation of flow rate. To the user, it will look like everything PR 2/3 toward #1323!

## changelog

- create FlowRateField component, remove stub
- RadioOption component can take children (for inline display)
- mess with StepForm column widths

## review requests

-  All step forms: transfer, consolidate, distribute, mix, heck even pause, should look OK in terms of alignment (@howisthisnamenottakenyet please give everything a glance)
- User can change flow rate for aspirate and dispense on all steps (except pause) -- but in this PR, they don't get saved to `commands` part of the file. Next PR!
- Flow rate modal interaction/validation works as described below:

1. When you open flow rate modal, the existing selection will be shown
2. When you type something into the Custom field, the Custom radio button will get selected automatically if it wasn't already
3. The only error you should see is about too many digits, until you click Done
4. Clicking Done when you exceed max volume or have 0 or no number will "dirty" the field. Now the min/max caption becomes an error. But the too many digits error has priority over the min/max out-of-bounds error.
5. At any time, selecting "Default" radio button will opt out of validation. Who cares what number or non-number is in that field if you're not using that field? Not us.

Phew, that was verbose.

### CODE REVIEW

File: protocol-designer/src/components/StepEditForm/FlowRateField/index.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  label?: $PropertyType<Props, 'label'>
+}
+
+type DP = {
+  updateValue: $PropertyType<Props, 'updateValue'>
+}
+
+type SP = $Diff<Props, DP>
+
+// Add a key to force re-constructing component when values change
```
Comment (IanLondon) [inline protocol-designer/src/components/StepEditForm/FlowRateField/index.js]: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key

File: protocol-designer/src/components/StepEditForm/MixForm.js [hunk 2]
```diff
@@ -64,9 +64,22 @@ const MixForm = (props: MixFormProps): React.Element<React.Fragment> => {
...
           <TipPositionField />
         </div>
+        <div className={styles.right_settings_column}>
+          <FlowRateField
+            name='aspirate_flowRate'
+            label='Aspirate Flow Rate'
+            pipetteFieldName='pipette'
+            flowRateType='aspirate'
+          />
+        </div>
```
Comment (IanLondon) [inline protocol-designer/src/components/StepEditForm/MixForm.js]: oops there should not be 2 right_settings_column's, this is a rebase artifact. Will fix!

File: protocol-designer/src/components/StepEditForm/TransferLikeForm.js [hunk 3]
```diff
@@ -121,7 +125,11 @@ const TransferLikeForm = (props: TransferLikeFormProps) => {
           </div>
           <div className={styles.right_settings_column}>
```
Comment (IanLondon) [inline protocol-designer/src/components/StepEditForm/TransferLikeForm.js]: Again I got 2 right_settings_column's in the rebase, will fix

File: protocol-designer/src/components/StepEditForm/FlowRateField/FlowRateField.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+type State = {
+  showModal: boolean,
+  modalFlowRate: ?string,
+  modalUseDefault: boolean,
+  pristine: boolean
+}
+
+export default class FlowRateField extends React.Component<Props, State> {
+  constructor (props: Props) {
```
Comment (b-cooper) [inline protocol-designer/src/components/StepEditForm/FlowRateField/FlowRateField.js]: It looks like the custom value is being maintained even if you switch your pipette. (e.g. select p10, set flow rate to custom 300 µL/s, select p300).  In a discussion with @howisthisnamenottakenyet regarding similar behavior in Tip Positioning, we agreed that an expected behavior would be to fallback to the default if the pipette changes.  You could achieve this with something along the lines of ...
```
componentDidUpdate(prevProps){ 
  if(prevProps.pipetteSelection !== this.props.pipetteSelection) {
    this.setState({modalUseDefault: true}) 
  } 
}`

Comment (IanLondon) [reply protocol-designer/src/components/StepEditForm/FlowRateField/FlowRateField.js]: Good catch!

Comment (IanLondon) [reply protocol-designer/src/components/StepEditForm/FlowRateField/FlowRateField.js]: The componentDidUpdate example would visually update the field, but the unsavedForm would still have the last flow rate value. That value just wouldn't be shown. I think I'd have to make the change in `handleFormChange.js`

Comment (IanLondon) [reply protocol-designer/src/components/StepEditForm/FlowRateField/FlowRateField.js]: Follow-up PR for this: #2154

**Pull request merged**